filters/filter.go

```diff
@@ -22,103 +22,171 @@ package filters
 
 import (
 	"bytes"
+	"errors"
 	"fmt"
-	"sort"
+	"strings"
 )
 
-// logicalOp is a logical operator
-type logicalOp string
+var (
+	errInvalidFilterPattern = errors.New("invalid filter pattern defined")
+)
+
+// LogicalOp is a logical operator
+type LogicalOp string
 
 // A list of supported logical operators
 const (
-	conjunction logicalOp = "&&"
-	disjunction logicalOp = "||"
+	// Conjunction is logical AND
+	Conjunction LogicalOp = "&&"
+	// Disjunction is logical OR
+	Disjunction LogicalOp = "||"
+	// WildcardChar is the wildcard character
+	WildcardChar = "*"
 )
 
-// IDFilter matches an id against certain conditions
-type IDFilter interface {
+// Filter matches a string against certain conditions
+type Filter interface {
 	fmt.Stringer
 
 	// Matches returns true if the conditions are met
-	Matches(id string) bool
+	Matches(val string) bool
+}
+
+// NewFilter supports startsWith, endsWith, contains and a single wildcard
+// TODO(martinm): add rest of glob matching support and negation
+func NewFilter(pattern string) (Filter, error) {
+	idx := strings.Index(pattern, WildcardChar)
+
+	if idx == -1 {
+		// No wildcards
+		return newEqualityFilter(pattern), nil
+	}
+
+	if len(pattern) == 1 {
+		// Whole thing is wildcard
+		return newAllowFilter(), nil
+	}
+
+	if idx == len(pattern)-1 {
+		// Wildcard at end
+		return newStartsWithFilter(pattern[:len(pattern)-1]), nil
+	}
+
+	secondIdx := strings.Index(pattern[idx+1:], WildcardChar)
+	if secondIdx == -1 {
+		if idx == 0 {
+			return newEndsWithFilter(pattern[1:]), nil
+		}
+
+		return newMultiFilter([]Filter{
+			newStartsWithFilter(pattern[:idx]),
+			newEndsWithFilter(pattern[idx+1:]),
+		}, Conjunction), nil
+	}
+
+	if idx != 0 || secondIdx != len(pattern)-2 || secondIdx == 0 {
```

Comment:
Sure, will simplify.

Yeah I thought about the ** case, I made it an error right now because while we can support it, you should easily just convert it to *. It's also weird if we support "**", but not "***"